cli: add userfile ls CLI command to list uploaded files

adityamaru opened the pull request:
This change adds support for a `userfile ls` command which allows users
to list file(s) which have been uploaded to the user-scoped
FileTableStorage.

The ls command accepts a single, optional CLI arg which can either be
a well-formed userfile URI or a glob pattern.

The semantics of userfile ls are as follows:
`userfile ls` : List all files in the default user FileTableStorage
tables `defaultdb.public.userfiles_$USERNAME`

`userfile ls userfile://db.schema.tableprefix/glob/pattern` : List all
files which match the glob pattern and are in the
`db.schema.tableprefix` tables.

`userfile ls glob/pattern` : List files which match the glob pattern in
the default user FileTableStorage tables
`defaultdb.public.userfiles_$USERNAME`

Informs: #51222

Release note (cli change): `userfile` now supports an `ls` command which
allows users to list the files they have uploaded to the user-scoped
FileTableStorage. `userfile ls` accepts a single, optional CLI arg which
can either be a well-formed userfile URI or a glob pattern.
The latter defaults to searching in the default
FileTableStorage tables `defaultdb.public.userfiles_$USRENAME`.

adityamaru committed:
cli: add userfile ls CLI command to list uploaded files

This change adds support for a `userfile ls` command which allows users
to list file(s) which have been uploaded to the user-scoped
FileTableStorage.

The ls command accepts a single, optional CLI arg which can either be
a well-formed userfile URI or a glob pattern.

The semantics of userfile ls are as follows:
`userfile ls` : List all files in the default user FileTableStorage
tables `defaultdb.public.userfiles_$USERNAME`

`userfile ls userfile://db.schema.tableprefix/glob/pattern` : List all
files which match the glob pattern and are in the
`db.schema.tableprefix` tables.

`userfile ls glob/pattern` : List files which match the glob pattern in
the default user FileTableStorage tables
`defaultdb.public.userfiles_$USERNAME`

Release note (cli change): `userfile` now supports an `ls` command which
allows users to list the files they have uploaded to the user-scoped
FileTableStorage. `userfile ls` accepts a single, optional CLI arg which
can either be a well-formed userfile URI or a glob pattern.
The latter defaults to searching in the default
FileTableStorage tables `defaultdb.public.userfiles_$USRENAME`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51482)
<!-- Reviewable:end -->

knz reviewed (commented):
nit: `public` in the last line of the release note

> Reviewed 10 of 12 files at r1, 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-:-MCLr1DC1nmSxiHRLNSu:b1b5udi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @miretskiy)

___
*[pkg/cli/cli_test.go, line 1414 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCLpNnD59z7x_lVMpkS:-MCLpNnE2YjB-0gLhsNr:b-5o35v4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d26187fa9decfb7215b88e4c7f0f21c6c977ee76/pkg/cli/cli_test.go#L1414)):*
> ```Go
> 
>   nodelocal         upload and delete nodelocal files
>   userfile          upload and delete user scoped files
> ```

I wonder if now is not the time to fold "nodelocal" and "userfile" under a single "bulk" group command?
___
*[pkg/cli/nodelocal.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCLpXdD5Qwn6NlGcwKj:-MCLpXdD5Qwn6NlGcwKk:biw42bk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d26187fa9decfb7215b88e4c7f0f21c6c977ee76/pkg/cli/nodelocal.go#L56)):*
> ```Go
> 	defer reader.Close()
> 
> 	ctx, cancel := context.WithCancel(context.Background())
> ```

Is there a need for this cancel function? I don't see an async process being spawned by uploadFile below.
___
*[pkg/cli/userfile.go, line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCLpbyN2Ch4zxiaBGhb:-MCLpbyN2Ch4zxiaBGhc:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/d26187fa9decfb7215b88e4c7f0f21c6c977ee76/pkg/cli/userfile.go#L55)):*
> ```Go
> 
> 	ctx, cancel := context.WithCancel(context.Background())
> 	defer cancel()
> ```

ditto
___
*[pkg/cli/userfile.go, line 68 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCLpksg5-9wql19e28N:-MCLpksg5-9wql19e28O:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/3807979fb640886f8c241975cf253bd91dfc528d/pkg/cli/userfile.go#L68)):*
> ```Go
> 
> 	ctx, cancel := context.WithCancel(context.Background())
> 	defer cancel()
> ```

ditto
___
*[pkg/sql/copy_file_upload_test.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCLpzVB2ds26y4SsoKW:-MCLpzVC43f-6qL_MGLt:b-6jr5ys) ([raw file](https://github.com/cockroachdb/cockroach/blob/3807979fb640886f8c241975cf253bd91dfc528d/pkg/sql/copy_file_upload_test.go#L58)):*
> ```Go
> 		testSendFile = strings.TrimPrefix(testSendFile, "/")
> 		uri = fmt.Sprintf("nodelocal://self/%s", testSendFile)
> 	} else if copyInternalTable == UserFileUploadTable {
> ```

I might have mentioned a switch here before
___
*[pkg/storage/cloudimpl/filetable/file_table_read_writer.go, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCLrG9aD6yuQ0-TwW-e:-MCLrG9aD6yuQ0-TwW-f:b-7zzhu4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3807979fb640886f8c241975cf253bd91dfc528d/pkg/storage/cloudimpl/filetable/file_table_read_writer.go#L42)):*
> ```Go
> type FileToTableSystem interface {
> 	FileSize(ctx context.Context, filename string) (int64, error)
> 	ListFiles(ctx context.Context, pattern string) ([]string, error)
> ```

May I suggest the iterator pattern here (have the function return an object with a Next and Close methods).
We have been bitten multiple times in the past with commands failing with an OOM error because listings were too large.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51482)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @knz, and @miretskiy)

___
*[pkg/cli/cli_test.go, line 1414 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCLpNnD59z7x_lVMpkS:-MCMYM0X7paGY-3eNNCI:bo21t1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d26187fa9decfb7215b88e4c7f0f21c6c977ee76/pkg/cli/cli_test.go#L1414)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I wonder if now is not the time to fold "nodelocal" and "userfile" under a single "bulk" group command?
</blockquote></details>

We discussed this a bit last week and tossed around the idea of a `cockroach io upload local-file <external-URI>` that could handle any URI instead of making commands specific to `userfile` and `nodelocal`, etc, but our conclusion was instead that we want to simply deprecate the `nodelocal upload` command in favor of `userfile` and all new functionality (`ls`, `rm`, etc) will be solely under `userfile`.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51482)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @miretskiy)

___
*[pkg/cli/cli_test.go, line 1414 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCLpNnD59z7x_lVMpkS:-MCMhnHN5QdrjfVea5XP:bxr9yfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d26187fa9decfb7215b88e4c7f0f21c6c977ee76/pkg/cli/cli_test.go#L1414)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

We discussed this a bit last week and tossed around the idea of a `cockroach io upload local-file <external-URI>` that could handle any URI instead of making commands specific to `userfile` and `nodelocal`, etc, but our conclusion was instead that we want to simply deprecate the `nodelocal upload` command in favor of `userfile` and all new functionality (`ls`, `rm`, etc) will be solely under `userfile`.
</blockquote></details>

Cool :+1: 


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51482)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @knz, and @miretskiy)

___
*[pkg/storage/cloudimpl/filetable/file_table_read_writer.go, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCLrG9aD6yuQ0-TwW-e:-MCMiOOo7zOLMg5aDr37:beyxpd5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3807979fb640886f8c241975cf253bd91dfc528d/pkg/storage/cloudimpl/filetable/file_table_read_writer.go#L42)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

May I suggest the iterator pattern here (have the function return an object with a Next and Close methods).
We have been bitten multiple times in the past with commands failing with an OOM error because listings were too large.
</blockquote></details>

This is a good point, but this interface comes from ExternalStorage so we should fix it there first.
___
*[pkg/storage/cloudimpl/filetable/file_table_read_writer.go, line 643 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCMij2j106L3ANBEPbn:-MCMij2j106L3ANBEPbo:b548yht) ([raw file](https://github.com/cockroachdb/cockroach/blob/3807979fb640886f8c241975cf253bd91dfc528d/pkg/storage/cloudimpl/filetable/file_table_read_writer.go#L643)):*
> ```Go
> ) ([]string, error) {
> 	var files []string
> 	listFilesQuery := fmt.Sprintf(`SELECT filename FROM %s WHERE filename LIKE '%s' ORDER BY
> ```

this is an unfortunate amount of copy/paste duplication, just to _run_ the same query using an internal executor vs an pg db conn one. I assume you already tried to have only one FileToTableSystem impl with a pluggable query-runner?  


<!-- Sent from Reviewable.io -->

adityamaru reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51482)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, and @miretskiy)

___
*[pkg/cli/nodelocal.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCLpXdD5Qwn6NlGcwKj:-MCNrQY8AVsIrLQko5Gf:b-7kzs3g) ([raw file](https://github.com/cockroachdb/cockroach/blob/d26187fa9decfb7215b88e4c7f0f21c6c977ee76/pkg/cli/nodelocal.go#L56)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Is there a need for this cancel function? I don't see an async process being spawned by uploadFile below.
</blockquote></details>

fixed in a previous PR, rebased.
___
*[pkg/cli/userfile.go, line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCLpbyN2Ch4zxiaBGhb:-MCNrd5B003WcrJc5R9z:bnlb34k) ([raw file](https://github.com/cockroachdb/cockroach/blob/d26187fa9decfb7215b88e4c7f0f21c6c977ee76/pkg/cli/userfile.go#L55)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

fixed in the previous PR, rebased.
___
*[pkg/cli/userfile.go, line 68 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCLpksg5-9wql19e28N:-MCNrTUwDqyunANXSx9X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3807979fb640886f8c241975cf253bd91dfc528d/pkg/cli/userfile.go#L68)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[pkg/sql/copy_file_upload_test.go, line 58 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCLpzVB2ds26y4SsoKW:-MCNrjPRAR0Jmv-gvHkc:bnlb34k) ([raw file](https://github.com/cockroachdb/cockroach/blob/3807979fb640886f8c241975cf253bd91dfc528d/pkg/sql/copy_file_upload_test.go#L58)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I might have mentioned a switch here before
</blockquote></details>

fixed in the previous PR, rebased.
___
*[pkg/storage/cloudimpl/filetable/file_table_read_writer.go, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCLrG9aD6yuQ0-TwW-e:-MCNrlPM8TJtfnGjSMvP:bridm73) ([raw file](https://github.com/cockroachdb/cockroach/blob/3807979fb640886f8c241975cf253bd91dfc528d/pkg/storage/cloudimpl/filetable/file_table_read_writer.go#L42)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

This is a good point, but this interface comes from ExternalStorage so we should fix it there first.
</blockquote></details>

Yup, a follow up to change the semantics for all the ExternalStorage will be a better place to do this. I've created an issue - #51517
___
*[pkg/storage/cloudimpl/filetable/file_table_read_writer.go, line 643 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCMij2j106L3ANBEPbn:-MCSHaqJ57wsKaHD5ur3:bq9o86i) ([raw file](https://github.com/cockroachdb/cockroach/blob/3807979fb640886f8c241975cf253bd91dfc528d/pkg/storage/cloudimpl/filetable/file_table_read_writer.go#L643)):*
> Yeah
agreed. refactored to introduce an Executor interface that is implemented by an "internal" and "network" backed executor.


<!-- Sent from Reviewable.io -->

adityamaru commented:
This should be RFAL, I refactored the code to introduce two flavors of the executor used by the FileToTableSystem to interact with the user scoped tables. One backed by a network conn (used by CLI) and the other backed by an internal SQL conn. 

I also switched some of the queries to bound parameters from Sprintf().

dt commented on pkg/cli/userfile.go:
```diff
@@ -69,8 +104,117 @@ func openUserFile(source string) (io.ReadCloser, error) {
+	// User has not specified a destination URI/path. We use the default URI
+	// scheme and host, and the basename from the source arg as the path.
+	if destination == "" {
+		sourceFilename := filepath.Base(source)
```
this should be `path` not `filepath` since it is a host-independent network supplied path.

dt commented on pkg/storage/cloudimpl/file_table_storage.go:
```diff
@@ -95,13 +125,34 @@ func (f *fileTableStorage) Conf() roachpb.ExternalStorage {
+		return "", errors.Newf("basename %s changes to %s on normalization. "+
+			"userfile does not permit such constructs.", basename, path.Clean(basename))
+	}
+	return filepath.Join(prefix, basename), nil
```
s/filepath/path/

adityamaru reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51482)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @knz, and @miretskiy)

___
*[pkg/cli/userfile.go, line 112 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCJKYsyiZjfRzpnrpDM-r5-112:-MChaxPZ1rRbSAQsMBX9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0712d0b5d9969f48237dcfa7c11bcccdb83f5c69/pkg/cli/userfile.go#L112)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

this should be `path` not `filepath` since it is a host-independent network supplied path.
</blockquote></details>

Done.
___
*[pkg/storage/cloudimpl/file_table_storage.go, line 142 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51482#-MCJKYsyiZjfRzpnrpDU-r5-142:-MChayJNAe-XnurXiXsF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0712d0b5d9969f48237dcfa7c11bcccdb83f5c69/pkg/storage/cloudimpl/file_table_storage.go#L142)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

s/filepath/path/
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

adityamaru commented:
TFTR!

bors r+

adityamaru commented:
bors r+
